File: 0129_sum_root_to_leaf/sum_numbers.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Time: O(n)
 * Space: O(1)
 */
func sumNumbers(root *TreeNode) int {
	return dfs(root, 0)
}

func dfs(node *TreeNode, curr int) int {
	if node != nil {
		curr = curr*10 + node.Val
		if node.Left == nil && node.Right == nil {
			return curr
		}
		return dfs(node.Left, curr) + dfs(node.Right, curr)
	}
	return 0
}

func main() {

}

File: 0129_sum_root_to_leaf/sum_numbers.py
class TreeNode:
    def __init__(self, val=0, left=None,
                 right=None):
        self.val = val
        self.left = left
        self.right = right


def sum_numbers(root: TreeNode) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    return dfs(root, 0)


def dfs(node: TreeNode, curr: int) -> int:
    if node:
        curr = curr * 10 + node.val
        if not node.left and not node.right:
            return curr
        return dfs(node.left, curr) + dfs(node.right, curr)
    return 0
